plone-5/instance/src/ukstats.csv_type/src/ukstats/csv_type/adapter.py

```diff
@@ -0,0 +1,50 @@
+""" behavior module """
+
+import logging
+
+from io import StringIO
+import csv
+from plone.memoize import ram
+from zope.component import adapter
+from zope.interface import implementer
+from zope.publisher.interfaces.browser import IBrowserRequest
+
+from eea.api.dataconnector.interfaces import IDataProvider
+from ukstats.csv_type.content.csv_type import ICSVMarker
+
+logger = logging.getLogger(__name__)
+
+
+@adapter(ICSVMarker, IBrowserRequest)
+@implementer(IDataProvider)
+class DataProviderForCsvws(object):
+    """ data provider for connectors """
+
+    def __init__(self, context, request):
+        self.context = context
+        self.request = request
+
+    # TO DO: persistent caching, periodical refresh, etc
+    @ram.cache(lambda func, self: (self.context.modified(), self.request.form))
+    def _provided_data(self):
+        """ provided data """
+        print(self.context.csv)
```

Comment:
Is this needed?

Reply:
The print can go.

The caching probably wants improving; this is just taken from one of the EEA adapters, but from an outdated version. Leave it for now and fix in a separate issue?

Reply:
Yeah cool, its not the end of the world.